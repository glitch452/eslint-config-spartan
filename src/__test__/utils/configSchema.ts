import { z } from 'zod';

const INITIAL_ECMA_VERSION = 3;
const INITIAL_ECMA_YEAR = 2015;
const NEXT_YEAR = new Date().getFullYear() + 1;
const MIN_ECMA_VERSION = 5;
const MAX_ECMA_VERSION = NEXT_YEAR - INITIAL_ECMA_YEAR + MIN_ECMA_VERSION + 1;

const ecmaVersionSchema = z.union([
  z.literal('latest'),
  z.literal(INITIAL_ECMA_VERSION),
  z.number().int().min(MIN_ECMA_VERSION).max(MAX_ECMA_VERSION),
  z.number().int().min(INITIAL_ECMA_YEAR).max(NEXT_YEAR),
]);
const sourceTypeSchema = z.enum(['script', 'module', 'commonjs']);
const globalConfSchema = z.boolean().or(z.enum(['off', 'readable', 'readonly', 'writable', 'writeable']));

const MIN_SEVERITY = 0;
const MAX_SEVERITY = 2;
const severitySchema = z.number().int().min(MIN_SEVERITY).max(MAX_SEVERITY);
const stringSeveritySchema = z.enum(['off', 'warn', 'error']);

const objectMetaPropertiesSchema = z.object({
  /** @deprecated Use `meta.name` instead. */
  name: z.string().optional(),
  /** @deprecated Use `meta.version` instead. */
  version: z.string().optional(),
  meta: z.object({ name: z.string().optional(), version: z.string().optional() }).optional(),
});

const esTreeParserSchema = z.union([
  objectMetaPropertiesSchema.extend({ parse: z.function() }),
  objectMetaPropertiesSchema.extend({ parseForESLint: z.function() }),
]);

const processorSchema = objectMetaPropertiesSchema.extend({
  supportsAutofix: z.boolean().optional(),
  preprocess: z.function(),
  postprocess: z.function(),
});

const parserOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    ecmaFeatures: z
      .object({
        globalReturn: z.boolean(),
        impliedStrict: z.boolean(),
        jsx: z.boolean(),
        experimentalObjectRestSpread: z.boolean(),
      })
      .partial()
      .catchall(z.any()),
  })
  .partial()
  .catchall(z.any());

const languageOptionsSchema = z
  .object({
    ecmaVersion: ecmaVersionSchema,
    sourceType: sourceTypeSchema,
    globals: z.record(globalConfSchema),
    parser: esTreeParserSchema,
    parserOptions: parserOptionsSchema,
  })
  .partial();

const linterOptionsSchema = z
  .object({
    noInlineConfig: z.boolean().optional(),
    reportUnusedDisableDirectives: z.union([severitySchema, stringSeveritySchema, z.boolean()]).optional(),
  })
  .optional();

const externalSpecifierSchema = z.object({
  name: z.string().optional(),
  url: z.string().optional(),
});

const replacedBySchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  plugin: externalSpecifierSchema.optional(),
  rule: externalSpecifierSchema.optional(),
});

const deprecatedInfoSchema = z.object({
  message: z.string().optional(),
  url: z.string().optional(),
  replacedBy: replacedBySchema.array().optional(),
  deprecatedSince: z.string().optional(),
  availableUntil: z.string().nullable().optional(),
});

const ruleMetaDataSchema = z
  .object({
    docs: z
      .object({
        description: z.string(),
        category: z.string(),
        // Added z.string() to accommodate storybook plugin
        // Added z.record(z.unknown()) to accommodate typescript-eslint plugin
        recommended: z.union([z.boolean(), z.string(), z.record(z.unknown())]),
        url: z.string(),
      })
      .partial(),
    messages: z.record(z.string()),
    // Added z.null() added to accommodate the cypress plugin
    fixable: z.enum(['code', 'whitespace']).nullable(),
    schema: z.unknown(),
    deprecated: z.boolean().or(deprecatedInfoSchema),
    replacedBy: z.string().array().readonly(),
    // Added 'error' to accommodate the security plugin
    type: z.enum(['problem', 'suggestion', 'layout', 'error']),
    hasSuggestions: z.boolean(),
  })
  .partial();

const ruleModuleSchema = z.object({
  create: z.function(),
  meta: ruleMetaDataSchema.optional(),
});

const pluginSchema = z
  .object({
    configs: z.record(z.unknown()),
    environments: z.record(
      z.object({ globals: z.record(globalConfSchema), parserOptions: parserOptionsSchema }).partial(),
    ),
    processors: z.record(processorSchema),
    rules: z.record(ruleModuleSchema),
  })
  .partial();

export const configSchema = z
  .object({
    name: z.string(),
    files: z.array(z.string().or(z.string().array())),
    ignores: z.string().array(),
    languageOptions: languageOptionsSchema,
    linterOptions: linterOptionsSchema,
    processor: z.string().or(processorSchema),
    plugins: z.record(pluginSchema),
    rules: z.unknown(),
    settings: z.record(z.unknown()),
  })
  .partial();
