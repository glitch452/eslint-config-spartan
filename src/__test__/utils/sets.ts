type Widen<T> = T | (TSReset.WidenLiteral<T> & {});

/**
 * `A âˆ© B`.
 * Compute the intersection of unique elements contained in both {@link a} and {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function intersection<A, B>(a: Widen<A>[], b: Widen<B>[]): Extract<Widen<A>, Widen<B>>[] {
  const setB = new Set<Widen<A> | Widen<B>>(b);
  return [...new Set<Widen<A>>(a)].filter((x): x is Extract<Widen<A>, Widen<B>> => setB.has(x));
}

/**
 * `A - B` a.k.a. `A / B`.
 * Compute the unique elements in {@link a} that do not also exist in {@link b}.
 * @param a
 * @param b
 * @returns
 */
export function difference<A, B>(a: A[], b: B[]): Widen<A>[] {
  const setB = new Set<A | B>(b);
  return [...new Set<A>(a)].filter((x): x is A => !setB.has(x));
}
