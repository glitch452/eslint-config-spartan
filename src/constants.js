export const configNamePrefix = 'eslint-config-spartan';

export const prefixes = /** @type {const} */ ({
  angular: '@angular-eslint',
  angularTemplate: '@angular-eslint/template',
  chaiExpect: 'chai-expect',
  chaiFriendly: 'chai-friendly',
  cypress: 'cypress',
  i18next: 'i18next',
  i18nJson: 'i18n-json',
  import: 'import',
  jest: 'jest',
  jestDom: 'jest-dom',
  jsDoc: 'jsdoc',
  jsxA11y: 'jsx-a11y',
  mdx: 'mdx',
  mocha: 'mocha',
  next: '@next/next',
  playwright: 'playwright',
  promise: 'promise',
  react: 'react',
  reactHooks: 'react-hooks',
  regExp: 'regexp',
  security: 'security',
  storybook: 'storybook',
  stylistic: '@stylistic',
  tailwindCss: 'tailwindcss',
  tanStackQuery: '@tanstack/query',
  testingLibrary: 'testing-library',
  typescriptEnum: 'typescript-enum',
  typescriptEsLint: '@typescript-eslint',
  unicorn: 'unicorn',
  unusedImports: 'unused-imports',
  vitest: 'vitest',
});
