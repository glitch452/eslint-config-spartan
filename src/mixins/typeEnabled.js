import { configNamePrefix } from '../constants.js';
import { files } from '../utils/index.js';
import typescriptEsLint from 'typescript-eslint';
/** @import { Linter } from 'eslint' */
/** @import { ConfigWithExtends } from 'typescript-eslint' */

/** @typedef {NonNullable<ConfigWithExtends['languageOptions']>['parserOptions']} ConfigParserOptions */

const tsRulesIndex = 2;
const recommendedTypeCheckedRules = typescriptEsLint.configs.recommendedTypeCheckedOnly[tsRulesIndex].rules;

/**
 * The `typeEnabled` mixin creates an eslint config with type-enabled rules for
 * [typescript-eslint](https://typescript-eslint.io). It requires an external config to enable the `@typescript-eslint`
 * plugin itself since it only includes rules. i.e. it's meant to bs used with the base config created with `buildConfig`. It also disables base eslint rules that are replaced with the equivalent
 * typescript-eslint rules.
 *
 * > [!NOTE]
 * >
 * > Note: This config also adds an ignores property to prevent these rules from being applied to code blocks within
 * > markdown files when the `mdx` mixin is being used since the parser cannot access that code.
 *
 * Rule Prefixes:
 *
 * - The `typescript-eslint` rules are prefixed with `@typescript-eslint`
 * @param {object} options
 * @param {Linter.Config['files']=} options.files Set the files for this config. By default, this applies to all `js/ts` variant file extensions.
 * @param {Linter.Config['ignores']=} options.ignores Set the ignores for this config. By default, this applies to all `.md/*` and `.mdx/*` to ignore code within markdown file code blocks.
 * @param {ConfigParserOptions=} options.parserOptions See <https://typescript-eslint.io/packages/parser/#configuration> for more information on the ParserOptions available.
 * @returns {Linter.Config}
 */
export function typeEnabled(options = {}) {
  return {
    name: `${configNamePrefix}/${typeEnabled.name}`,
    files: options.files ?? [files.jsTs],
    ignores: options.ignores ?? [`${files.mdMdx}/**/*`],
    languageOptions: { parserOptions: options.parserOptions },
    rules: {
      ...recommendedTypeCheckedRules,

      // Disable JS rules that have TS versions below
      'dot-notation': 'off',

      '@typescript-eslint/dot-notation': 'error',
      '@typescript-eslint/no-misused-promises': ['error', { checksVoidReturn: false }],
      '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error',
      '@typescript-eslint/no-unnecessary-condition': 'error',
      '@typescript-eslint/no-unsafe-argument': 'off',
      '@typescript-eslint/no-unsafe-assignment': 'off',
      '@typescript-eslint/no-unsafe-call': 'off',
      '@typescript-eslint/no-unsafe-member-access': 'off',
      '@typescript-eslint/no-unsafe-return': 'off',
      '@typescript-eslint/prefer-includes': 'error',
      '@typescript-eslint/prefer-nullish-coalescing': 'error',
      '@typescript-eslint/prefer-optional-chain': 'error',
      '@typescript-eslint/prefer-reduce-type-parameter': 'error',
      '@typescript-eslint/switch-exhaustiveness-check': 'error',
      '@typescript-eslint/unbound-method': ['error', { ignoreStatic: true }],
    },
  };
}
