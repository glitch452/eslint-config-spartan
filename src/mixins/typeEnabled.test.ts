import typescriptEsLint from 'typescript-eslint';
import { typeEnabled } from './typeEnabled.js';
import { configSchema } from '../__test__/utils/configSchema.js';
import { difference, intersection } from '../__test__/utils/sets.js';
import { files } from '../utils/index.js';
import { CONFIG_NAME_PREFIX, prefixes } from '../constants.js';
import {
  getDeprecatedRules,
  getEnabledRules,
  getTypeInfoRequiredRules,
  getWarnRules,
  listRules,
} from '../__test__/utils/rules.js';
import { Linter } from 'eslint';

describe(typeEnabled.name, () => {
  const config = typeEnabled();

  const validRules = [
    ...new Linter({ configType: 'eslintrc' }).getRules().keys(),
    ...listRules((typescriptEsLint.plugin as any).rules, prefixes.typescriptEsLint),
  ];
  const deprecatedRules = [
    ...getDeprecatedRules(new Linter({ configType: 'eslintrc' }).getRules()),
    ...getDeprecatedRules((typescriptEsLint.plugin as any).rules, prefixes.typescriptEsLint),
  ];
  const typeInfoRequiredRules = getTypeInfoRequiredRules(
    (typescriptEsLint.plugin as any).rules,
    prefixes.typescriptEsLint,
  );

  it('should create a valid eslint config', () => {
    const actual = () => configSchema.parse(config);
    expect(actual).not.toThrow();
  });

  it('should add the expected plugins to the config', () => {
    const actual = Object.keys(config.plugins ?? {}).sort();
    const expected = [].sort();
    expect(actual).toStrictEqual(expected);
  });

  it('should configure the expected rules', () => {
    expect(config.rules).toMatchSnapshot();
  });

  it('should track the list of type information required rules', () => {
    expect(typeInfoRequiredRules).toMatchSnapshot();
  });

  it('should set the name value', () => {
    expect(config.name).toBe(`${CONFIG_NAME_PREFIX}/${typeEnabled.name}`);
  });

  it('should set the default files value', () => {
    const actual = config.files;
    const expected = [files.jsTs];
    expect(actual).toStrictEqual(expected);
  });

  it('should set the given files value', () => {
    const filesValue = ['files.txt'];
    const actual = typeEnabled({ files: filesValue }).files;
    expect(actual).toStrictEqual(filesValue);
  });

  it('should set the default ignores value', () => {
    const actual = config.ignores;
    const expected = [`${files.mdMdx}/**/*`];
    expect(actual).toStrictEqual(expected);
  });

  it('should set the given ignores value', () => {
    const ignoresValue = ['ignores.txt'];
    const actual = typeEnabled({ ignores: ignoresValue }).ignores;
    expect(actual).toStrictEqual(ignoresValue);
  });

  it('should set the given parserOptions value', () => {
    const options = { parserOptions: { ecmaVersion: 'latest' } } as const;
    const actual = typeEnabled(options).languageOptions;
    expect(actual).toStrictEqual(options);
  });

  it('should only configure rules that exist', () => {
    const configuredRules = Object.keys(config.rules ?? {});
    const actual = difference(configuredRules, validRules);
    expect(actual).toStrictEqual([]);
  });

  it('should not configure rules that are deprecated', () => {
    const configuredAndEnabledRules = getEnabledRules(config.rules);
    const actual = intersection(configuredAndEnabledRules, deprecatedRules);
    expect(actual).toStrictEqual([]);
  });

  it('should not configure any rules as warn', () => {
    const configuredAndWarnRules = getWarnRules(config.rules);
    expect(configuredAndWarnRules).toStrictEqual([]);
  });

  it('should only configure rules that require type information', () => {
    const configuredTsEsLintRules = Object.keys(config.rules ?? {}).filter((x) =>
      x.startsWith(prefixes.typescriptEsLint),
    );
    const actual = difference(configuredTsEsLintRules, typeInfoRequiredRules);
    expect(actual).toStrictEqual([]);
  });
});
