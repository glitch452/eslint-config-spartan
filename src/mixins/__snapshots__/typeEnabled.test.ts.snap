// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typeEnabled > should configure the expected rules 1`] = `
{
  "@typescript-eslint/await-thenable": "error",
  "@typescript-eslint/dot-notation": "error",
  "@typescript-eslint/no-base-to-string": "error",
  "@typescript-eslint/no-duplicate-type-constituents": "error",
  "@typescript-eslint/no-floating-promises": "error",
  "@typescript-eslint/no-for-in-array": "error",
  "@typescript-eslint/no-implied-eval": "error",
  "@typescript-eslint/no-misused-promises": [
    "error",
    {
      "checksVoidReturn": false,
    },
  ],
  "@typescript-eslint/no-redundant-type-constituents": "error",
  "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error",
  "@typescript-eslint/no-unnecessary-condition": "error",
  "@typescript-eslint/no-unnecessary-type-assertion": "error",
  "@typescript-eslint/no-unsafe-argument": "off",
  "@typescript-eslint/no-unsafe-assignment": "off",
  "@typescript-eslint/no-unsafe-call": "off",
  "@typescript-eslint/no-unsafe-enum-comparison": "error",
  "@typescript-eslint/no-unsafe-member-access": "off",
  "@typescript-eslint/no-unsafe-return": "off",
  "@typescript-eslint/prefer-includes": "error",
  "@typescript-eslint/prefer-nullish-coalescing": "error",
  "@typescript-eslint/prefer-optional-chain": "error",
  "@typescript-eslint/prefer-reduce-type-parameter": "error",
  "@typescript-eslint/require-await": "error",
  "@typescript-eslint/restrict-plus-operands": "error",
  "@typescript-eslint/restrict-template-expressions": "error",
  "@typescript-eslint/switch-exhaustiveness-check": [
    "error",
    {
      "requireDefaultForNonUnion": true,
    },
  ],
  "@typescript-eslint/unbound-method": [
    "error",
    {
      "ignoreStatic": true,
    },
  ],
  "default-case": "off",
  "dot-notation": "off",
  "no-implied-eval": "off",
  "require-await": "off",
}
`;

exports[`typeEnabled > should track the list of type information required rules 1`] = `
[
  "@typescript-eslint/await-thenable",
  "@typescript-eslint/consistent-return",
  "@typescript-eslint/consistent-type-exports",
  "@typescript-eslint/dot-notation",
  "@typescript-eslint/naming-convention",
  "@typescript-eslint/no-array-delete",
  "@typescript-eslint/no-base-to-string",
  "@typescript-eslint/no-confusing-void-expression",
  "@typescript-eslint/no-duplicate-type-constituents",
  "@typescript-eslint/no-floating-promises",
  "@typescript-eslint/no-for-in-array",
  "@typescript-eslint/no-implied-eval",
  "@typescript-eslint/no-meaningless-void-operator",
  "@typescript-eslint/no-misused-promises",
  "@typescript-eslint/no-mixed-enums",
  "@typescript-eslint/no-redundant-type-constituents",
  "@typescript-eslint/no-throw-literal",
  "@typescript-eslint/no-unnecessary-boolean-literal-compare",
  "@typescript-eslint/no-unnecessary-condition",
  "@typescript-eslint/no-unnecessary-qualifier",
  "@typescript-eslint/no-unnecessary-template-expression",
  "@typescript-eslint/no-unnecessary-type-arguments",
  "@typescript-eslint/no-unnecessary-type-assertion",
  "@typescript-eslint/no-unnecessary-type-parameters",
  "@typescript-eslint/no-unsafe-argument",
  "@typescript-eslint/no-unsafe-assignment",
  "@typescript-eslint/no-unsafe-call",
  "@typescript-eslint/no-unsafe-enum-comparison",
  "@typescript-eslint/no-unsafe-member-access",
  "@typescript-eslint/no-unsafe-return",
  "@typescript-eslint/no-unsafe-unary-minus",
  "@typescript-eslint/no-useless-template-literals",
  "@typescript-eslint/non-nullable-type-assertion-style",
  "@typescript-eslint/only-throw-error",
  "@typescript-eslint/prefer-destructuring",
  "@typescript-eslint/prefer-find",
  "@typescript-eslint/prefer-includes",
  "@typescript-eslint/prefer-nullish-coalescing",
  "@typescript-eslint/prefer-optional-chain",
  "@typescript-eslint/prefer-promise-reject-errors",
  "@typescript-eslint/prefer-readonly",
  "@typescript-eslint/prefer-readonly-parameter-types",
  "@typescript-eslint/prefer-reduce-type-parameter",
  "@typescript-eslint/prefer-regexp-exec",
  "@typescript-eslint/prefer-return-this-type",
  "@typescript-eslint/prefer-string-starts-ends-with",
  "@typescript-eslint/promise-function-async",
  "@typescript-eslint/require-array-sort-compare",
  "@typescript-eslint/require-await",
  "@typescript-eslint/restrict-plus-operands",
  "@typescript-eslint/restrict-template-expressions",
  "@typescript-eslint/return-await",
  "@typescript-eslint/strict-boolean-expressions",
  "@typescript-eslint/switch-exhaustiveness-check",
  "@typescript-eslint/unbound-method",
  "@typescript-eslint/use-unknown-in-catch-callback-variable",
]
`;
